src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,142 @@
+    """
+    def __init__(self, *args, **kwargs):
+        super(MDSAutoscaler, self).__init__(*args, **kwargs)
+        self.set_mgr(self)
+        self.fs_map: Optional[dict] = None
+        self.old_notify_type = "NONE"
+
+    def get_service(self, fs_name) -> Optional[List[orchestrator.ServiceDescription]]:
+        result = None
+        try:
+            #service = "mds.{}".format(fs_name)
+            service = "mds.newb"
+            args = ()
+            completion = self.describe_service(service_type='mds',
+                                               service_name=service,
+                                               refresh=True)
+            self._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+            result = completion.result
+        except Exception as e:
+            self.log.exception("{}: exception while fetching service info"
+                               .format(e))
+            pass
+        self.log.info("service info:{}".format(result[0]))
+        return result
+
+    def update_daemon_count(self, fs_name, abscount) -> Optional[ServiceSpec]:
+        svclist = self.get_service(fs_name)
+        if svclist is None or len(svclist) == 0:
+            self.log.warn("Failed to fetch MDS service details for {}"
+                           .format(fs_name))
+            return None
+
+        svc = svclist[0]
+        self.log.info("newsvc:{}".format(svc))
+        self.log.info("newsvc.spec:{}".format(svc.spec))
+        self.log.info("new count:{}".format(abscount))
+
+        hosts = []
+        for h in svc.spec.placement.hosts:
+            hosts.append(h.hostname)
+        ps = PlacementSpec(hosts=hosts, count=abscount)
+        newsvc = ServiceSpec(service_type=svc.spec.service_type,
+                             service_id=svc.spec.service_id,
+                             placement=ps)
+        return newsvc
+
+    def spawn_mds(self, fs_name, count):
+        newsvc = self.update_daemon_count(fs_name, count+1)
+        if newsvc is None:
+            return
+
+        completion = self.apply_mds(newsvc)
+        self._orchestrator_wait([completion])
+        orchestrator.raise_if_exception(completion)
+
+    def kill_mds(self, fs_name, count):
+        newsvc = self.update_daemon_count(fs_name, count-1)
+        if newsvc is None:
+            return
+
+        completion = self.apply_mds(newsvc)
+        self._orchestrator_wait([completion])
+        orchestrator.raise_if_exception(completion)
+
+    def get_required_standby_count(self):
+        total = 1
+        assert self.fs_map is not None
+        for fs in self.fs_map.get('filesystems'):
+            self.log.info("getting standby_count_wanted")
+            total = max(total, fs.get('mdsmap').get('standby_count_wanted'))
+        return total
+
+    def get_current_standby_count(self):
+        # standbys are not grouped by filesystems in fs_map
+        # available = standby_replay + standby_active
```

Comment:
If ceph adm manages only MDS daemons that it creates, you can identify the standby daemons meant for a fileysystem by looking at `join_fscid` propetry of the standby daemons since the MDS daemons created by ceph adm are configured with `mds_join_fs` option.  What do you think?

Reply:
I think having the `mds_autoscaler` play nice with daemons launched outside of cephadm/orch is future work. The `get_daemons` method already returns in the daemon names for the MDSs that the orchestrator spawned so I don't think we need to get fancy (yet) by looking at `join_fscid`.